pkg/querier/metadata_handler.go

```diff
@@ -24,35 +31,79 @@ const (
 	statusError   = "error"
 )
 
-type metadataResult struct {
+type metadataSuccessResult struct {
 	Status string                      `json:"status"`
-	Data   map[string][]metricMetadata `json:"data,omitempty"`
-	Error  string                      `json:"error,omitempty"`
+	Data   map[string][]metricMetadata `json:"data"`
+}
+
+type metadataErrorResult struct {
+	Status string `json:"status"`
+	Error  string `json:"error"`
 }
 
 // MetadataHandler returns metric metadata held by Cortex for a given tenant.
 // It is kept and returned as a set.
 func MetadataHandler(m MetadataQuerier) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		resp, err := m.MetricsMetadata(r.Context())
+		limit, err := validateLimits("limit", w, r)
+		if err != nil {
+			return
+		}
+
+		limitPerMetric, err := validateLimits("limit_per_metric", w, r)
+		if err != nil {
+			return
+		}
+
+		metric := r.FormValue("metric")
+		req := &client.MetricsMetadataRequest{
+			Limit:          int64(limit),
+			LimitPerMetric: int64(limitPerMetric),
+			Metric:         metric,
+		}
```

Comment:
If limit is set to 0, do we still need to send the request to Ingester? There is no result anyway